### ISSUE TITLE
Neovim/Template goto definition returns error: index out of range

### ISSUE THREAD

**Issue by catgoose** [a-h/templ#911]
**Before you begin**
```
❯ templ version
v0.2.778
```

**Describe the bug**
Attempting to use Neovim templ LSP to go to the definition of a templ template inside of a .templ file will produce error:

```lua
Error executing vim.schedule lua callback: /usr/local/share/nvim/runtime/lua/vim/lsp/util.lua:17
98: index out of range
stack traceback:
        [C]: in function '_str_byteindex_enc'
        /usr/local/share/nvim/runtime/lua/vim/lsp/util.lua:1798: in function 'locations_to_items
'
        /usr/local/share/nvim/runtime/lua/vim/lsp/handlers.lua:436: in function 'handler'
        /usr/local/share/nvim/runtime/lua/vim/lsp/client.lua:687: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
Press ENTER or type command to continue
```

**To Reproduce**
https://github.com/catgoose/templ_issue_repro

Follow steps in readme

**Expected behavior**
Issuing `vim.lsp.buf.definition` on a templ template reference in a templ file should go to that templates definition

**Screenshots**
![image](https://github.com/user-attachments/assets/47ee1abf-51b0-4f79-b602-f3e3e1d0c995)

Taken from the reproduction repo above

**Logs**
Log file `templ.log` included in repro repo

**`templ info` output**
```bash
❯ templ info
(✓) os [ goos=linux goarch=amd64 ]
(✓) go [ location=/usr/local/go/bin/go version=go version go1.23.1 linux/amd64 ]
(✓) gopls [ location=/home/jtye/go/bin/gopls version=golang.org/x/tools/gopls v0.16.2 ]
(✓) templ [ location=/home/jtye/go/bin/templ version=v0.2.778 ]
```

**Desktop (please complete the following information):**
 - OS: Ubuntu 24.04.1
 - templ version: v0.2.778
- Go version: go version go1.23.1 linux/amd64
- `gopls` version: golang.org/x/tools/gopls v0.16.2

**Additional context**
This seems to have been fixed here https://github.com/a-h/templ/issues/893 but the issue persists
Reported here https://github.com/catgoose/templ-goto-definition/issues/1

**Comment by a-h** [comment]
Thanks for providing a good reproduction. I had to run `:set filetype=templ` to get the LSP to attach to the buffer, but once I'd done that, I was able to type the default leader followed by gd to trigger the error (`\gd`).

I got similar logs as to the one in the repo. These are the important lines.

```
{"level":"info","ts":"2024-09-09T08:33:38+01:00","caller":"proxy/server.go:541","msg":"client -> server: Definition"}
{"level":"info","ts":"2024-09-09T08:33:38+01:00","caller":"proxy/server.go:71","msg":"updatePosition: found","uri":"file:///Users/adrian/github.com/catgoose/templ_issue_repro/views/index.templ","fromTempl":"6:5","toGo":"33:24"}
{"level":"info","ts":"2024-09-09T08:33:38+01:00","caller":"proxy/server.go:564","msg":"client -> server: Definition end"}
```

In particular this:

```json
{
   "level":"info",
   "ts":"2024-09-09T08:33:38+01:00",
   "caller":"proxy/server.go:71",
   "msg":"updatePosition: found",
   "uri":"file:///Users/adrian/github.com/catgoose/templ_issue_repro/views/index.templ",
   "fromTempl":"6:5",
   "toGo":"33:24"
}
```

Note that templ believes that line 6, col 5 in the templ file maps to the line 33, col 24 in the Go code.

I can verify this in the templ LSP web server. Note the highlighted (yellow) section.

<img width="1920" alt="Screenshot 2024-09-09 at 08 33 12" src="https://github.com/user-attachments/assets/0e1ff36a-6144-4b8d-8654-4dae7a2c07c9">

This looks correct.

I had updated my version of gopls as per your reproduction instructions. When I put it back, the problem went away, so I went back through the history of gopls to find when the error started happening.

If I install gopls v0.15.2 (`go install golang.org/x/tools/gopls@v0.15.2`), it works OK, but if I use v0.15.3 or greater (`go install golang.org/x/tools/gopls@v0.15.3`), then I get the error.

I'm not sure what changed between those versions. Maybe a reshape of the JSON structures to support a later version of the LSP spec? But, progress on reproduction.

**Comment by a-h** [comment]
Comparison of gopls: https://github.com/golang/tools/compare/gopls/v0.15.2...gopls/v0.15.3

**Comment by catgoose** [comment]
What is strange is such as with #893 if you open the file for the target definition, then go to definition works. 

Target buffer not open, error presented:

```json
{
  "level": "info",
  "ts": "2024-09-09T06:43:13-05:00",
  "caller": "proxy/server.go:541",
  "msg": "client -> server: Definition"
}
{
  "level": "info",
  "ts": "2024-09-09T06:43:13-05:00",
  "caller": "proxy/server.go:71",
  "msg": "updatePosition: found",
  "uri": "file:///home/catgoose/git/templ_issue_repro/views/index.templ",
  "fromTempl": "6:4",
  "toGo": "33:23"
}
{
  "level": "info",
  "ts": "2024-09-09T06:43:13-05:00",
  "caller": "proxy/server.go:564",
  "msg": "client -> server: Definition end"
}
```

Opened target buffer:

```json
{
  "level": "info",
  "ts": "2024-09-09T06:45:19-05:00",
  "caller": "proxy/server.go:650",
  "msg": "client -> server: DidOpen",
  "uri": "file:///home/catgoose/git/templ_issue_repro/views/content.templ"
}
{
  "level": "info",
  "ts": "2024-09-09T06:45:19-05:00",
  "caller": "proxy/server.go:674",
  "msg": "setting source map cache contents",
  "uri": "file:///home/catgoose/git/templ_issue_repro/views/content.templ"
}
{
  "level": "info",
  "ts": "2024-09-09T06:45:19-05:00",
  "caller": "proxy/server.go:681",
  "msg": "client -> server: DidOpen end"
}
{
  "level": "info",
  "ts": "2024-09-09T06:45:19-05:00",
  "caller": "proxy/client.go:52",
  "msg": "client <- server: PublishDiagnostics"
}
```

```json
{
  "level": "info",
  "ts": "2024-09-09T06:46:34-05:00",
  "caller": "proxy/server.go:541",
  "msg": "client -> server: Definition"
}
{
  "level": "info",
  "ts": "2024-09-09T06:46:34-05:00",
  "caller": "proxy/server.go:71",
  "msg": "updatePosition: found",
  "uri": "file:///home/catgoose/git/templ_issue_repro/views/index.templ",
  "fromTempl": "6:4",
  "toGo": "33:23"
}
{
  "level": "info",
  "ts": "2024-09-09T06:46:35-05:00",
  "caller": "proxy/server.go:564",
  "msg": "client -> server: Definition end"
}
```

So it seems to me that gopls is trying to target the wrong position, but if the buffer is open it catches it self and says "that's not the right line, but I can see it's on this line because I'm attached to that buffer so we will return that position"

**Comment by a-h** [comment]
I just noticed that!

The issue is that the gopls server is telling templ that the definition exists in a Go file that templ hasn't got in its `sourceMap` that maps from templ positions to Go code positions, so it's returning a position in the file that might not exist instead.

The issue is in the `convertGoRangeToTemplRange` function. If the Go range is in a `*_templ.go` file that it doesn't know about, it probably needs to load the associated `*.templ` file, and populate the source map for that file, then everything should work correctly.

However, I'm not sure if we can rely on loading the file from the URI... one way to find out...

**Comment by tris203** [comment]
It will no longer error in Neovim nighly (as of last night, after this PR https://github.com/neovim/neovim/pull/30288)
And it won't error in Stable after the next release as the fix got backported.

However, it still maps to the wrong spot if the source map doesn't exist. 
I see where the problem is in ```convertGoRangeToTemplRange```.
Happy to take a look shortly if you want me to @a-h

**Comment by axzilla** [comment]
> It will no longer error in Neovim nighly (as of last night, after this PR [neovim/neovim#30288](https://github.com/neovim/neovim/pull/30288)) And it won't error in Stable after the next release as the fix got backported.
> 
> However, it still maps to the wrong spot if the source map doesn't exist. I see where the problem is in `convertGoRangeToTemplRange`. Happy to take a look shortly if you want me to @a-h

Yes, I can confirm, the problem disappeared with the nightly version!

**Comment by a-h** [comment]
You went deep on this one @tris203 - well done! 😁

**Comment by tris203** [comment]
> You went deep on this one @tris203 - well done! 😁 

Thanks. The error is gone in neovim
But it will still you to the wrong line in the file until #912 is complete

**Comment by catgoose** [comment]
@a-h Could we get a release made with these changes?  Thanks!
